### DISCUSSION TITLE
Modbus treadmil

### DISCUSSION THREAD

**Discussion by carozoynarizota** [cagnulein/qdomyos-zwift#2253] [Q&A]
Hi, I installed the app on a Raspberry pi 3b+ just to see how it goes. All of what a read within the guides or youtube makes me think that the app act just as a bridge between zwift and bluetooth enabled devices. These can be FTMS treadmils or BLE sensors.

What I am trying to figure out if it will help me control my AC motor  through the VFD using modbus. Despite being a modbus branch, I see no option to set for example how the Pi should connect to an external device through an RS485 adapter.

My setup is just this basic treadmill controlled by buttons that goes directly to the VFD. To know my pace I put a hall sensor and with an ESP32 over BT I have that on zwift. Setting different speeds or intervale training is a nightmare with the buttons, as every push is a 0.1Hz step in the AC motor.

**Comment by cagnulein** [comment]
hi @carozoynarizota the mod is branch has a static dictionary for a modbus device. if you need a specific modbus dictionary share the specs of your device

let me know

**Reply by carozoynarizota** [reply]
Hi Roberto, my AC motor controller is a VFD-M model. Documentation can be found by Delta or Danfoss as its brand, model 176F7300.

On the attached pdf, appendix D there is detailed information about modbus communication.
[Danfoss VLT Micro - 176F7330.pdf](https://github.com/cagnulein/qdomyos-zwift/files/14782286/Danfoss.VLT.Micro.-.176F7330.pdf)

For modbus communications I been doing tests with this settings:
00: MODBUS ASCII mode, <7,N,2>
03: MODBUS RTU mode, <8,N,2>

With manufacturer specific software I was able to read all the registers (and get the query also). Some of them are to control specific parameters, like setting the AC motor frequency, and at specific addresses toggling certain bits I remember you can for example start, stop or reverse the motor.

With all this information I was able to run with an arduino a code that start the treadmill and make it increase speed every certain time. Of course code is not mine but copied from the internet, just with a few modifications to match the documentation. For example no one uses on their VFDs those parity settings, most of them has modes that are supported on software serial library and for mine it has to be hardware serial.

Well thats all I can say now, I don't know if its usefull or if it justifies the effort to get you involved.

As final examples I put here the output I get through reading all the registers in ASCII or RTU. I also have examples setting frequency, or running/stopping the treadmil. 
[2022-06-19 VFDC Read Parameter ASCII Log.txt](https://github.com/cagnulein/qdomyos-zwift/files/14782404/2022-06-19.VFDC.Read.Parameter.ASCII.Log.txt)
[2022-06-19 VFDC Read Parameter RTU Log.txt](https://github.com/cagnulein/qdomyos-zwift/files/14782405/2022-06-19.VFDC.Read.Parameter.RTU.Log.txt)

**Reply by cagnulein** [reply]
@carozoynarizota ok i did a first raw implementation. I just implemented the read mechaninms, let's see if everything works.

Are you on Windows right with the serial port?

I had hard coded the COM4 in the code and i'm reading the first 113 registers as you did. If it works you will see the value of each ones in the debug log. Let me know!

The build is here https://github.com/cagnulein/qdomyos-zwift/actions/runs/8523485903

**Reply by carozoynarizota** [reply]
Hi Roberto, I work all day on this! I think its gonna kill me.

Before a lot of trial an error I was able to install and run your software on two different devices:

RaspberryPi 4: It opens and start looking for BLE devices, seems to work OK.
Windows11 PC: It starts OK but doesn't seem to be opening the bluetooth

What I don't know is if a supposed to do something different than with the master branch, if I am not wrong thats what I am installing by following the documentation. What should I do to try this implementation?

For today I ran out of time....

**Reply by cagnulein** [reply]
first of all are you using this on windows? https://github.com/cagnulein/qdomyos-zwift/actions/runs/8523485903/artifacts/1377685293

just run it and collect a debug log with the debug log setting enabled

**Reply by carozoynarizota** [reply]
Roberto yes, now with the executable I use it in windows. From your previous message I didn't know that there was a way to end up with an executable program directly under windows. Will dig on that later.

For the testing I was using Windows 11, set my module to COM4 and run the program. It doesn't seem to be opening that port at all.

I can confirm through delta VFDSoft that RS485 module is working and with that I can Start/Stop the TM (ASCII 7,N,2):

[5/4/2024 21:46:58:19] Send > :010620000001D8\x0D\x0A .......  Receive > :010620000001D8\x0D\x0A (Stop)
[5/4/2024 21:46:51:42] Send > :010620000002D7\x0D\x0A .......  Receive > :010620000002D7\x0D\x0A (Run)

Of course I am not opening both programs at the same time.
[debug-vie__abr__5_21_50_07_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14891865/debug-vie__abr__5_21_50_07_2024.log)

Thats all for now....

**Reply by cagnulein** [reply]
sorry @carozoynarizota i forgot a thing

new build https://github.com/cagnulein/qdomyos-zwift/actions/runs/8595896304

**Reply by carozoynarizota** [reply]
Hi Roberto, how do I create the executable for this? From github? visual studio? some sort of script?

**Reply by cagnulein** [reply]
@carozoynarizota here https://github.com/cagnulein/qdomyos-zwift/actions/runs/8595896304/artifacts/1393246971

**Reply by carozoynarizota** [reply]
Now it opens the port and try to establish connection. But when something is wrong you may not have a reply but still have an error message on the display.

Unfortunately I wasn't really able to understand through the manual what each error means. If I set the VFD to 9600 ASCII 7,N,2 then the error is C-14, and for 9600 RTU 8,N,2 the error is C-03.

Run the program twice on each configuration and get the same error each one of them. The problem I think may come from this settings or by CRC computation.

[Error C03 (RTU8N2).log](https://github.com/cagnulein/qdomyos-zwift/files/14913037/Error.C03.RTU8N2.log)
[Error C14 (ASCII7N2).log](https://github.com/cagnulein/qdomyos-zwift/files/14913038/Error.C14.ASCII7N2.log)

**Reply by cagnulein** [reply]
@carozoynarizota thanks, i tried to simplify the code because i was reading all the 113 registers at once and maybe the motor doesn't handle it. I added also some more new logs.

I set the communication to RTU8N2 so you can use this one

https://github.com/cagnulein/qdomyos-zwift/actions/runs/8611535691

you can get the new version in the link at the bottom of the page (the msvc no python link)

Let me know!

**Reply by carozoynarizota** [reply]
Hi Roberto, with this release the display looks like flickering for a while....something like a few seconds maybe. Then it ends witch a C-14 error.

About the log I don't see the registers read but just a line about modbus connected. I don't know if it is regarding some effective reading. I don't see there the commands sent nor the replies.

[debug-mar__abr__9_08_30_44_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14926303/debug-mar__abr__9_08_30_44_2024.log)
[debug-mar__abr__9_08_32_53_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/14926305/debug-mar__abr__9_08_32_53_2024.log)

**Reply by cagnulein** [reply]
yes it's working, check the lines with "handleReadResponse". they are our values! ok so i will try to implement a start and setting the speed now

**Reply by cagnulein** [reply]
@carozoynarizota if you have a log or a quick description to start and set speed I can quickly implement it instead of trying ot understand the datasheet (if you already tried this it will be quicker!)
Let me know

**Reply by carozoynarizota** [reply]
Yes! Have a rest and give me some time tu put that in order.

I have the start/stop commands and I think also a table setting frequency by 0.1Hz step.

Also the frequency speed relationship....but more on this later as I think its not linear really as it shoud be.

**Reply by carozoynarizota** [reply]
Roberto I've being doing some tests to get the right commands. I think we only need Start, Stop and Send Frequency.

Frequency is from 0 to 60, with two decimals but is sent in HEXA. So ie to set 24,44Hz, then you have to send 2444 as an integer. In HEXA 0988 will be our value.
Then to set a new frequency 01 for device ID,06 write operation, 2001 the expected address, an 0988 our value without CRC should be :010620010988

In the attached file you have the Start, Stop, and Freq commands from 0 to 20Hz. Both ASCII and RTU.

If it happens to be usefull to read frequency, then command is :010321020001
If a communication error occurs, you won't receive an answer or maybe an error response. The display will show you an error like "C-##", that error is cleared by the reset command also in the table.

Let me know if you need further details or tests.

[Lecturas.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/14968701/Lecturas.xlsx)

**Reply by cagnulein** [reply]
done @carozoynarizota  https://github.com/cagnulein/qdomyos-zwift/actions/runs/8687385936

What I did:
- start implemented
- stop implemented
- frequency changed implemented: i'm actually sending the km/h value as frequency. so you will have to tell me how to scale it (for example 1hz means 0.2 km/h)

Let's see if this will work!
Let me know!

**Reply by carozoynarizota** [reply]
Hi Roberto! Its looking good!

I see we are still reading the configuration registers. That gaves me some errors while reading, I thing some sort of timing through request it can't handle. I think there is no need now for it, we can just keep with the basic commands.

About speed there is a formula right, but it doesn't work quite well because of mechanic issues. As far as I know the AC motor turns driven by the VFD frequency, it cannot fall behind that without failure. So speed must be fixed to it, in Km/h the factor is 0,352918.

Formulas in here, green columns for frequency, speed (Km/h) and Pace (min/km)
[Centauro MX7.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/15003587/Centauro.MX7.xlsx)
ie 60hz * 0.352918 = 21,18Kmh

The problem is while I step in there is (I supose) some slipping on the driving belt, so real speed falls behind that I don't know by how much exactly. But what I really care is the measured speed by the front roll where the sensor is located.

Can you take that measure to adjust the frequency? You can see on the log file there is a BLE sensor detected as HamsperPOD. Thats what I am using that now for zwift......control is my finger :(

[debug-mar__abr__16_19_40_01_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/15003620/debug-mar__abr__16_19_40_01_2024.log)

One more thing, when play start on the app it just send start command, I think it should also send the speed setting on the screen. That way it will start at a very low speed on the very first time, then when paused I will resume later at the las speed used.

I think thats all for now.

**Reply by cagnulein** [reply]
that's great @carozoynarizota  !
https://github.com/cagnulein/qdomyos-zwift/actions/runs/8717986181

in this build i did:
- removed the reading of the parameters
- added a default value for the speed when you hit start
- added the conversion for the speed

What I suggest now is to use the speed gain and speed offset settings to calibrate it, it should be more or less linear, what do you think?
In the worst case, yes I can use an external sensor to trim it. IN the log I saw a Polar Run sensor, is that one?

**Reply by carozoynarizota** [reply]
Hi Roberto! It works OK! About the formula I see the following, when the app reaches 21,5Km I see on the VFD if am not wrong now 59.5Hz. If I push one more to set 22kmh I get error C-03, I think we´re sending a value bigger than 60 that is no allowed.

About the formula, for 21,5Kmh the Frequency should be 60,92Hz (21.5/0.352918). In fact if I am going to correct it through offset or gain or trim is not important, but we have to be shure no to send a value bigger than 60.00

About the default value of 0.8, its very slow, It would be better something like 5Km/h. With someone on top of the TM  motor current gets too high at very low speeds

**Reply by cagnulein** [reply]
done both! https://github.com/cagnulein/qdomyos-zwift/actions/runs/8738653869

**Reply by carozoynarizota** [reply]
Then about the sensor, de Polar one I don't even know where it is at home! It should be sleeping. LOL
It's very inaccurate, non linear. And specially a pain in the ass to switch between shoes.

BTLE device:  "227315067511546" "Polar RUN 34FB542E"

The one I use is this one, and ESP32 with just a hall sensor attached to the front roller:

BTLE device:  "263920670644630" "HamsterPod" Num UUIDs 1

You can see both on the following log:

[debug-mié__abr__17_22_53_55_2024.log](https://github.com/cagnulein/qdomyos-zwift/files/15025465/debug-mie__abr__17_22_53_55_2024.log)

By now I was just taking time to check the things regarding the control of the VFD, didn't have time to dig more onto the app.

The offset and gain configurations are under settings right?

Speed management is made by speed on discrete jumps of 0.5kmh, thats a lot. Is always this way or just in the demo?
Does QZ allow to modify by pace?

I am testing this on PC, but later I will be using and orange pi if it works. Can I run two instances of QZ one on Windows/Linux and other on Android/IOS as a front end?

**Reply by cagnulein** [reply]
> The offset and gain configurations are under settings right?

yes advanced settings

> Speed management is made by speed on discrete jumps of 0.5kmh, thats a lot.

you can customize this in the treadmill section, treadmill step setting.

> Does QZ allow to modify by pace?

sure!

> I am testing this on PC, but later I will be using and orange pi if it works. Can I run two instances of QZ one on Windows/Linux and other on Android/IOS as a front end?

never tested this, but you can use VNC for example or the webview already builtin in qz!

why don't connect the serial port directly on the android one otherwise?

**Reply by carozoynarizota** [reply]
> why don't connect the serial port directly on the android one otherwise?

Didn't though on this!. If my phone supports OTG this would be an option!

I think that for a future release I must put some effort on get this all through the ESP32 and expose everything through it as an FTMS to QZ.

Does QZ make any difference on a tablet? Does it support big screens? For example Garmin connect shows itself just like in the phone....horrible.

**Reply by cagnulein** [reply]
yes an esp32 could be a good solution :)

qz is adapting on the tablet and also with the zoom ui you can customize the size of the tiles. and also you can customize the tiles that you want to see! everything is possible with QZ!

**Reply by carozoynarizota** [reply]
Hi Roberto! I tried late at night last changes. They work perfect, de 5kmh default speed and the top setting for the frequency not to put the VFD under error message.

I also found how to change the step from 0.5 to 0.1, and even the predefined speeds...thats awsome. Still didn´t get where to change from speed to pace.

I thinks its almost done. What´s next? What about the fixed com4 on window? What does it take to pick the right port in Linux?

**Reply by cagnulein** [reply]
> Hi Roberto! I tried late at night last changes. They work perfect, de 5kmh default speed and the top setting for the frequency not to put the VFD under error message.
> 
> 

Amazing!

> 
> I also found how to change the step from 0.5 to 0.1, and even the predefined speeds...thats awsome. 

Super!

> Still didn´t get where to change from speed to pace.

You have to enable the pace tile in the settings. Did you?

> 
> 
> 
> I thinks its almost done. What´s next? What about the fixed com4 on window? What does it take to pick the right port in Linux?

I can add a new setting for the comport. The Linux port should be already work, just need to change the com4 label :)

**Reply by carozoynarizota** [reply]
Hi Roberto.

> You have to enable the pace tile in the settings. Did you?

Yes it is enabled, but the +/- are just on the speed tile. I don't know if it supports to place changes directly from the pace tile.

I also having a lot of problems with the tiles. Tried to arrange them but I only move one and then it goes back to its position in line, I am also not being able to si if I can change its size. There's also a font problem being bigger than the tile.

![Screenshot 2024-04-21 113033](https://github.com/cagnulein/qdomyos-zwift/assets/31801337/70cb010f-a850-431a-a600-31755af4daa3)

> I can add a new setting for the comport. The Linux port should be already work, just need to change the com4 label :)

The links you gave me have only the APK for android and the windows versions, not the linux one. I don't know how to install this demo version for linux.

I installed again the actual release following the steps in the installation manual, under Ubuntu for windows. It runs OK, but as it doesn't find any TM I am not able to see if the tiles arrangement works there. I set the Fake Treadmill feature on, restart QZ but nothing changes (no dashboard).

Moving to something different. On the final release I see you publish the android and ios version on each store. But it seems to be also a link for a windows version on google, but its seems to be unavailable:

![Screenshot 2024-04-21 134511](https://github.com/cagnulein/qdomyos-zwift/assets/31801337/327e25fd-9851-408a-bd42-59796c55d24a)

**Reply by cagnulein** [reply]
> Hi Roberto.
> 
> 
> 
> > You have to enable the pace tile in the settings. Did you?
> 
> 
> 
> Yes it is enabled, but the +/- are just on the speed tile. I don't know if it supports to place changes directly from the pace tile.

you have to change the speed from the speed tile, but since pace and speed are linked you will see pace changes too

> 
> 
> 
> I also having a lot of problems with the tiles. Tried to arrange them but I only move one and then it goes back to its position in line, I am also not being able to si if I can change its size. There's also a font problem being bigger than the tile.

for the font issue you have to change the dpi for qz. you can do in the advanced settings of the executable on windows. for the order of the tiles you have to change the order index in the tiles settings.

> 
> 
> 
> ![Screenshot 2024-04-21 113033](https://github.com/cagnulein/qdomyos-zwift/assets/31801337/70cb010f-a850-431a-a600-31755af4daa3)
> 
> 
> 
> > I can add a new setting for the comport. The Linux port should be already work, just need to change the com4 label :)
> 
> 
> 
> The links you gave me have only the APK for android and the windows versions, not the linux one. I don't know how to install this demo version for linux.

i'm not building the linux version on the CI here on github. i can do it if you need.

> 
> 
> 
> I installed again the actual release following the steps in the installation manual, under Ubuntu for windows. It runs OK, but as it doesn't find any TM I am not able to see if the tiles arrangement works there. I set the Fake Treadmill feature on, restart QZ but nothing changes (no dashboard).

no don't follow this, qz runs natively on linux. 

> 
> 
> 
> Moving to something different. On the final release I see you publish the android and ios version on each store. But it seems to be also a link for a windows version on google, but its seems to be unavailable:
> 

i'm not updating the windows version on the microsoft store because of the microsoft policies. the windows version is only here on github

> 
> 
> ![Screenshot 2024-04-21 134511](https://github.com/cagnulein/qdomyos-zwift/assets/31801337/327e25fd-9851-408a-bd42-59796c55d24a)
> 
> 
> 
> 
> 
> 
> 
> 
> 
>

**Reply by carozoynarizota** [reply]
> no don't follow this, qz runs natively on linux.

Roberto I tried to run the linux file picked from the following link. That was on ubuntu desktop under VirtualBOX, change it properties as an executable and run it but nothing happens. Am I doing it right?

https://github.com/cagnulein/qdomyos-zwift/actions/runs/8768731219

On the nighlty build there is no linux version built. Are all this features implemented there?

I will be ordering two OTG adapters from aliexpress (for android and ipad). Ipad would be my first option, but I would also like to have the linux option.

**Comment by cagnulein** [comment]
ok i will implement this in the next days!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 28 mar 2024 alle ore 01:18 carozoynarizota <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
amazing! i will implement this in the next days!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 13 apr 2024 alle 23:35 carozoynarizota <
***@***.***> ha scritto:
